numpy/core/src/multiarray/multiarraymodule.c

```diff
@@ -1577,25 +1577,37 @@ _array_fromobject_generic(
     if (PyArray_CheckExact(op) || (subok && PyArray_Check(op))) {
         oparr = (PyArrayObject *)op;
         if (type == NULL) {
-            if (!copy && STRIDING_OK(oparr, order)) {
+            if ((copy == NPY_COPY_IF_NEEDED || copy == NPY_COPY_NEVER) && 
+                STRIDING_OK(oparr, order)) {
                 ret = oparr;
                 Py_INCREF(ret);
                 goto finish;
             }
             else {
+                if( copy == NPY_COPY_NEVER ) {
+                    PyErr_SetString(PyExc_ValueError,
+                                    "Unable to avoid copy while creating a new array.");
+                    return NULL;
+                }
                 ret = (PyArrayObject *)PyArray_NewCopy(oparr, order);
                 goto finish;
             }
         }
         /* One more chance */
         oldtype = PyArray_DESCR(oparr);
         if (PyArray_EquivTypes(oldtype, type)) {
-            if (!copy && STRIDING_OK(oparr, order)) {
+            if ((copy == NPY_COPY_IF_NEEDED || copy == NPY_COPY_NEVER) && 
+                STRIDING_OK(oparr, order)) {
                 Py_INCREF(op);
                 ret = oparr;
                 goto finish;
             }
             else {
+                if( copy == NPY_COPY_NEVER ) {
+                    PyErr_SetString(PyExc_ValueError,
+                                    "Unable to avoid copy while creating a new array.");
+                    return NULL;
+                }
```

Comment:
Can you run the test suite after removing the check in line 1606, see which tests fail, and then restore the line and reproduce them with `copy = NPY_COPY_NEVER` ?

Reply:
I will do that. I am curious though. What’s the motivation behind doing this?

Reply:
The following tests fail on removing this whole check (I commented out the if-block).

```
FAILED numpy/core/tests/test_multiarray.py::TestArrayCreationCopyArgument::test_order_mismatch[C-F-arr1]
FAILED numpy/core/tests/test_multiarray.py::TestArrayCreationCopyArgument::test_order_mismatch[C-None-arr1]
FAILED numpy/core/tests/test_multiarray.py::TestArrayCreationCopyArgument::test_order_mismatch[F-C-arr1]
FAILED numpy/core/tests/test_multiarray.py::TestArrayCreationCopyArgument::test_order_mismatch[F-None-arr1]
FAILED numpy/core/tests/test_multiarray.py::TestArrayCreationCopyArgument::test_striding_not_ok
```

Reply:
Ah! got it. These are different lines than what I covered earlier. I will be back on Monday and fix this.